File: backend/src/models/Incidencia.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/sequelize');

const Incidencia = sequelize.define('Incidencia', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  titulo: {
    type: DataTypes.STRING(200),
    allowNull: false,
    validate: {
      notEmpty: { msg: 'El título no puede estar vacío' },
    },
  },
  descripcion: {
    type: DataTypes.TEXT,
    allowNull: false,
    validate: {
      notEmpty: { msg: 'La descripción no puede estar vacía' },
    },
  },
  categoria: {
    type: DataTypes.STRING(80),
    allowNull: true,
  },
  prioridad: {
    type: DataTypes.ENUM('BAJA', 'MEDIA', 'ALTA', 'CRITICA'),
    allowNull: true,
    defaultValue: 'MEDIA',
  },
  estatus: {
    type: DataTypes.ENUM('ABIERTA', 'EN_PROCESO', 'EN_ESPERA', 'RESUELTA', 'CERRADA'),
    allowNull: false,
    defaultValue: 'ABIERTA',
  },
  usuario_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'usuarios',
      key: 'id',
    },
  },
  tecnico_id: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: {
      model: 'usuarios',
      key: 'id',
    },
  },
  activo: {
    type: DataTypes.BOOLEAN,
    allowNull: false,
    defaultValue: true,
  },
  fecha_creacion: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW,
  },
  fecha_cierre: {
    type: DataTypes.DATE,
    allowNull: true,
  },
}, {
  tableName: 'incidencias',
  timestamps: true,
  underscored: true,
});

module.exports = Incidencia;

File: frontend/lib/data/models/incidencia.dart
import 'usuario.dart';
import 'incidencia_log.dart';

class Incidencia {
  final int id;
  final String titulo;
  final String descripcion;
  final String? categoria;
  final String? prioridad;
  final String estatus;
  final int usuarioId;
  final int? tecnicoId;
  final bool activo;
  final DateTime? fechaCreacion;
  final DateTime? fechaCierre;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final Usuario? usuario;
  final Usuario? tecnico;
  final List<IncidenciaLog>? logs;

  const Incidencia({
    required this.id,
    required this.titulo,
    required this.descripcion,
    this.categoria,
    this.prioridad,
    required this.estatus,
    required this.usuarioId,
    this.tecnicoId,
    required this.activo,
    this.fechaCreacion,
    this.fechaCierre,
    this.createdAt,
    this.updatedAt,
    this.usuario,
    this.tecnico,
    this.logs,
  });

  factory Incidencia.fromJson(Map<String, dynamic> json) {
    return Incidencia(
      id: json['id'] as int,
      titulo: json['titulo'] as String,
      descripcion: json['descripcion'] as String,
      categoria: json['categoria'] as String?,
      prioridad: json['prioridad'] as String?,
      estatus: json['estatus'] as String,
      usuarioId: json['usuario_id'] as int,
      tecnicoId: json['tecnico_id'] as int?,
      activo: json['activo'] as bool? ?? true,
      fechaCreacion: json['fecha_creacion'] != null
          ? DateTime.tryParse(json['fecha_creacion'] as String)
          : null,
      fechaCierre: json['fecha_cierre'] != null
          ? DateTime.tryParse(json['fecha_cierre'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
      usuario: json['usuario'] != null
          ? Usuario.fromJson(json['usuario'] as Map<String, dynamic>)
          : null,
      tecnico: json['tecnico'] != null
          ? Usuario.fromJson(json['tecnico'] as Map<String, dynamic>)
          : null,
      logs: json['logs'] != null
          ? (json['logs'] as List<dynamic>)
              .map((l) => IncidenciaLog.fromJson(l as Map<String, dynamic>))
              .toList()
          : null,
    );
  }
}
